Build failure with latest servant libraries

```
Preprocessing library for servant-rawm-0.2.0.2..
Building library for servant-rawm-0.2.0.2..
[1 of 6] Compiling Servant.RawM.Internal.API ( src/Servant/RawM/Internal/API.hs, dist/build/Servant/RawM/Internal/API.o )
[2 of 6] Compiling Servant.RawM.Internal.Client ( src/Servant/RawM/Internal/Client.hs, dist/build/Servant/RawM/Internal/Client.o )

src/Servant/RawM/Internal/Client.hs:26:1: error:
    Could not find module ‘Servant.Common.Req’
    Use -v to see a list of the files searched for.
   |
26 | import Servant.Common.Req (Req, performRequest)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -93,13 +93,13 @@ matrix:
   #  compiler: ": #stack 8.0.1"
   #  addons: {apt: {packages: [libgmp-dev]}}
 
-  - env: BUILD=stack ARGS="--resolver lts-9"
-    compiler: ": #stack 8.0.2"
-    addons: {apt: {packages: [libgmp-dev]}}
+  # - env: BUILD=stack ARGS="--resolver lts-9"
+  #   compiler: ": #stack 8.0.2"
+  #   addons: {apt: {packages: [libgmp-dev]}}
 
-  - env: BUILD=stack ARGS="--resolver lts-11"
-    compiler: ": #stack 8.2.2"
-    addons: {apt: {packages: [libgmp-dev]}}
+  # - env: BUILD=stack ARGS="--resolver lts-11"
+  #   compiler: ": #stack 8.2.2"
+  #   addons: {apt: {packages: [libgmp-dev]}}
 
   - env: BUILD=stack ARGS="--resolver lts-12"
     compiler: ": #stack 8.4.3"

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,11 +1,18 @@
+# This is the complex Travis configuration, which is intended for use
+# on open source libraries which need compatibility across multiple GHC
+# versions, must work with cabal-install, and should be
+# cross-platform. For more information and other options, see:
+#
+# https://docs.haskellstack.org/en/stable/travis_ci/
+#
 # Copy these contents into the root directory of your Github project in a file
 # named .travis.yml
 
 # Use new container infrastructure to enable caching
 sudo: false
 
-# Choose a lightweight base image; we provide our own build tools.
-language: c
+# Do not choose a language; we provide our own build tools.
+language: generic
 
 # Caching so the next build will be fast too.
 cache:
@@ -13,6 +20,7 @@ cache:
   - $HOME/.ghc
   - $HOME/.cabal
   - $HOME/.stack
+  - $TRAVIS_BUILD_DIR/.stack-work
 
 # The different configurations we want to test. We have BUILD=cabal which uses
 # cabal-install, and BUILD=stack which uses Stack. More documentation on each
@@ -25,69 +33,115 @@ cache:
 # matrix, you can use a line such as:
 #     addons: {apt: {packages: [libfcgi-dev,libgmp-dev]}}
 matrix:
-  # Make the build finish as soon as all of the non-"allowed_failures" builds
-  # are finished building.  Don't wait for the "allowed_failures" to finish.
   fast_finish: true
   include:
   # We grab the appropriate GHC and cabal-install versions from hvr's PPA. See:
-  # - env: BUILD=cabal GHCVER=7.4.2 CABALVER=1.16 HAPPYVER=1.19.5 ALEXVER=3.1.7
-  #   compiler: ": #GHC 7.4.2"
-  #   addons: {apt: {packages: [cabal-install-1.16,ghc-7.4.2,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
-
-  # - env: BUILD=cabal GHCVER=7.6.3 CABALVER=1.16 HAPPYVER=1.19.5 ALEXVER=3.1.7
-  #   compiler: ": #GHC 7.6.3"
-  #   addons: {apt: {packages: [cabal-install-1.16,ghc-7.6.3,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
-
-  # - env: BUILD=cabal GHCVER=7.8.4 CABALVER=1.18 HAPPYVER=1.19.5 ALEXVER=3.1.7
-  #   compiler: ": #GHC 7.8.4"
-  #   addons: {apt: {packages: [cabal-install-1.18,ghc-7.8.4,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
-
-  # - env: BUILD=cabal GHCVER=7.10.3 CABALVER=1.22 HAPPYVER=1.19.5 ALEXVER=3.1.7
-  #   compiler: ": #GHC 7.10.3"
-  #   addons: {apt: {packages: [cabal-install-1.22,ghc-7.10.3,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
-
+  # https://github.com/hvr/multi-ghc-travis
+  #- env: BUILD=cabal GHCVER=7.0.4 CABALVER=1.16 HAPPYVER=1.19.5 ALEXVER=3.1.7
+  #  compiler: ": #GHC 7.0.4"
+  #  addons: {apt: {packages: [cabal-install-1.16,ghc-7.0.4,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+  #- env: BUILD=cabal GHCVER=7.2.2 CABALVER=1.16 HAPPYVER=1.19.5 ALEXVER=3.1.7
+  #  compiler: ": #GHC 7.2.2"
+  #  addons: {apt: {packages: [cabal-install-1.16,ghc-7.2.2,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+  #- env: BUILD=cabal GHCVER=7.4.2 CABALVER=1.16 HAPPYVER=1.19.5 ALEXVER=3.1.7
+  #  compiler: ": #GHC 7.4.2"
+  #  addons: {apt: {packages: [cabal-install-1.16,ghc-7.4.2,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+  #- env: BUILD=cabal GHCVER=7.6.3 CABALVER=1.16 HAPPYVER=1.19.5 ALEXVER=3.1.7
+  #  compiler: ": #GHC 7.6.3"
+  #  addons: {apt: {packages: [cabal-install-1.16,ghc-7.6.3,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+  #- env: BUILD=cabal GHCVER=7.8.4 CABALVER=1.18 HAPPYVER=1.19.5 ALEXVER=3.1.7
+  #  compiler: ": #GHC 7.8.4"
+  #  addons: {apt: {packages: [cabal-install-1.18,ghc-7.8.4,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+  #- env: BUILD=cabal GHCVER=7.10.3 CABALVER=1.22 HAPPYVER=1.19.5 ALEXVER=3.1.7
+  #  compiler: ": #GHC 7.10.3"
+  #  addons: {apt: {packages: [cabal-install-1.22,ghc-7.10.3,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
   - env: BUILD=cabal GHCVER=8.0.2 CABALVER=1.24 HAPPYVER=1.19.5 ALEXVER=3.1.7
     compiler: ": #GHC 8.0.2"
     addons: {apt: {packages: [cabal-install-1.24,ghc-8.0.2,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+  - env: BUILD=cabal GHCVER=8.2.2 CABALVER=2.0 HAPPYVER=1.19.5 ALEXVER=3.1.7
+    compiler: ": #GHC 8.2.2"
+    addons: {apt: {packages: [cabal-install-2.0,ghc-8.2.2,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+  - env: BUILD=cabal GHCVER=8.4.3 CABALVER=2.2 HAPPYVER=1.19.5 ALEXVER=3.1.7
+    compiler: ": #GHC 8.4.3"
+    addons: {apt: {packages: [cabal-install-2.2,ghc-8.4.3,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
+
+  # Build with the newest GHC and cabal-install. This is an accepted failure,
+  # see below.
+  # - env: BUILD=cabal GHCVER=head  CABALVER=head HAPPYVER=1.19.5 ALEXVER=3.1.7
+  #   compiler: ": #GHC HEAD"
+  #   addons: {apt: {packages: [cabal-install-head,ghc-head,happy-1.19.5,alex-3.1.7], sources: [hvr-ghc]}}
 
   # The Stack builds. We can pass in arbitrary Stack arguments via the ARGS
   # variable, such as using --stack-yaml to point to a different file.
-  # - env: BUILD=stack ARGS="--resolver lts-6"
-  #   compiler: ": #stack 7.10.3 (lts-6)"
-  #   addons: {apt: {packages: [ghc-7.10.3,libgmp-dev], sources: [hvr-ghc]}}
+  - env: BUILD=stack ARGS=""
+    compiler: ": #stack default"
+    addons: {apt: {packages: [libgmp-dev]}}
 
-  # - env: BUILD=stack ARGS="--resolver lts-7"
-  #   compiler: ": #stack 8.0.1 (lts-7)"
-  #   addons: {apt: {packages: [ghc-8.0.1,libgmp-dev], sources: [hvr-ghc]}}
+  #- env: BUILD=stack ARGS="--resolver lts-2"
+  #  compiler: ": #stack 7.8.4"
+  #  addons: {apt: {packages: [libgmp-dev]}}
 
-  # - env: BUILD=stack ARGS="--resolver lts-8"
-  #   compiler: ": #stack 8.0.2 (lts-8)"
-  #   addons: {apt: {packages: [ghc-8.0.2,libgmp-dev], sources: [hvr-ghc]}}
+  #- env: BUILD=stack ARGS="--resolver lts-3"
+  #  compiler: ": #stack 7.10.2"
+  #  addons: {apt: {packages: [libgmp-dev]}}
+
+  #- env: BUILD=stack ARGS="--resolver lts-6"
+  #  compiler: ": #stack 7.10.3"
+  #  addons: {apt: {packages: [libgmp-dev]}}
+
+  #- env: BUILD=stack ARGS="--resolver lts-7"
+  #  compiler: ": #stack 8.0.1"
+  #  addons: {apt: {packages: [libgmp-dev]}}
 
   - env: BUILD=stack ARGS="--resolver lts-9"
-    compiler: ": #stack 8.0.2 (lts-9)"
-    addons: {apt: {packages: [ghc-8.0.2,libgmp-dev], sources: [hvr-ghc]}}
+    compiler: ": #stack 8.0.2"
+    addons: {apt: {packages: [libgmp-dev]}}
+
+  - env: BUILD=stack ARGS="--resolver lts-11"
+    compiler: ": #stack 8.2.2"
+    addons: {apt: {packages: [libgmp-dev]}}
+
+  - env: BUILD=stack ARGS="--resolver lts-12"
+    compiler: ": #stack 8.4.3"
+    addons: {apt: {packages: [libgmp-dev]}}
 
   # Nightly builds are allowed to fail
   - env: BUILD=stack ARGS="--resolver nightly"
     compiler: ": #stack nightly"
     addons: {apt: {packages: [libgmp-dev]}}
 
-  # Build on OS X in addition to Linux
-  # - env: BUILD=stack ARGS="--resolver lts-6"
-  #   compiler: ": #stack 7.10.3 (lts-6) osx"
-  #   os: osx
+  # Build on macOS in addition to Linux
+  - env: BUILD=stack ARGS=""
+    compiler: ": #stack default osx"
+    os: osx
 
-  # - env: BUILD=stack ARGS="--resolver lts-7"
-  #   compiler: ": #stack 8.0.1 (lts-7) osx"
-  #   os: osx
+  # Travis includes an macOS which is incompatible with GHC 7.8.4
+  #- env: BUILD=stack ARGS="--resolver lts-2"
+  #  compiler: ": #stack 7.8.4 osx"
+  #  os: osx
 
-  # - env: BUILD=stack ARGS="--resolver lts-8"
-  #   compiler: ": #stack 8.0.2 (lts-8) osx"
-  #   os: osx
+  #- env: BUILD=stack ARGS="--resolver lts-3"
+  #  compiler: ": #stack 7.10.2 osx"
+  #  os: osx
+
+  #- env: BUILD=stack ARGS="--resolver lts-6"
+  #  compiler: ": #stack 7.10.3 osx"
+  #  os: osx
+
+  #- env: BUILD=stack ARGS="--resolver lts-7"
+  #  compiler: ": #stack 8.0.1 osx"
+  #  os: osx
 
   - env: BUILD=stack ARGS="--resolver lts-9"
-    compiler: ": #stack 8.0.2 (lts-9) osx"
+    compiler: ": #stack 8.0.2 osx"
+    os: osx
+
+  - env: BUILD=stack ARGS="--resolver lts-11"
+    compiler: ": #stack 8.2.2 osx"
+    os: osx
+
+  - env: BUILD=stack ARGS="--resolver lts-12"
+    compiler: ": #stack 8.4.3 osx"
     os: osx
 
   - env: BUILD=stack ARGS="--resolver nightly"
@@ -95,6 +149,7 @@ matrix:
     os: osx
 
   allow_failures:
+  - env: BUILD=cabal GHCVER=head  CABALVER=head HAPPYVER=1.19.5 ALEXVER=3.1.7
   - env: BUILD=stack ARGS="--resolver nightly"
   - os: osx
 
@@ -112,9 +167,9 @@ before_install:
 - |
   if [ `uname` = "Darwin" ]
   then
-    curl --insecure -L https://www.stackage.org/stack/osx-x86_64 | tar xz --strip-components=1 --include '*/stack' -C ~/.local/bin
+    travis_retry curl --insecure -L https://get.haskellstack.org/stable/osx-x86_64.tar.gz | tar xz --strip-components=1 --include '*/stack' -C ~/.local/bin
   else
-    curl -L https://www.stackage.org/stack/linux-x86_64 | tar xz --wildcards --strip-components=1 -C ~/.local/bin '*/stack'
+    travis_retry curl -L https://get.haskellstack.org/stable/linux-x86_64.tar.gz | tar xz --wildcards --strip-components=1 -C ~/.local/bin '*/stack'
   fi
 
   # Use the more reliable S3 mirror of Hackage
@@ -122,40 +177,71 @@ before_install:
   echo 'remote-repo: hackage.haskell.org:http://hackage.fpcomplete.com/' > $HOME/.cabal/config
   echo 'remote-repo-cache: $HOME/.cabal/packages' >> $HOME/.cabal/config
 
-  if [ "$CABALVER" != "1.16" ]
-  then
-    echo 'jobs: $ncpus' >> $HOME/.cabal/config
-  fi
 
 install:
 - echo "$(ghc --version) [$(ghc --print-project-git-commit-id 2> /dev/null || echo '?')]"
 - if [ -f configure.ac ]; then autoreconf -i; fi
 - |
+  set -ex
   case "$BUILD" in
     stack)
-      stack --no-terminal --install-ghc $ARGS test --only-dependencies
+      # Add in extra-deps for older snapshots, as necessary
+      #
+      # This is disabled by default, as relying on the solver like this can
+      # make builds unreliable. Instead, if you have this situation, it's
+      # recommended that you maintain multiple stack-lts-X.yaml files.
+
+      #stack --no-terminal --install-ghc $ARGS test --bench --dry-run || ( \
+      #  stack --no-terminal $ARGS build cabal-install && \
+      #  stack --no-terminal $ARGS solver --update-config)
+
+      # Build the dependencies
+      stack --no-terminal --install-ghc $ARGS test --bench --only-dependencies
       ;;
     cabal)
       cabal --version
       travis_retry cabal update
-      cabal install --only-dependencies --enable-tests --enable-benchmarks --force-reinstalls --ghc-options=-O0 --reorder-goals --max-backjumps=-1 $CABALARGS
+
+      # Get the list of packages from the stack.yaml file. Note that
+      # this will also implicitly run hpack as necessary to generate
+      # the .cabal files needed by cabal-install.
+      PACKAGES=$(stack --install-ghc query locals | grep '^ *path' | sed 's@^ *path:@@')
+
+      cabal install --only-dependencies --enable-tests --enable-benchmarks --force-reinstalls --ghc-options=-O0 --reorder-goals --max-backjumps=-1 $CABALARGS $PACKAGES
       ;;
   esac
+  set +ex
 
 script:
 - |
+  set -ex
   case "$BUILD" in
     stack)
-      stack --no-terminal $ARGS test --haddock --no-haddock-deps --flag servant-rawm:buildexample
+      stack --no-terminal $ARGS test --bench --no-run-benchmarks --haddock --no-haddock-deps --flag servant-rawm:buildexample
       ;;
     cabal)
-      cabal configure --enable-tests --enable-benchmarks -v2 --ghc-options="-O0 -Werror" --flags="buildexample"
-      cabal build
-      cabal check || [ "$CABALVER" == "1.16" ]
-      cabal test
-      cabal sdist
-      cabal copy
-      SRC_TGZ=$(cabal info . | awk '{print $2;exit}').tar.gz && \
-        (cd dist && cabal install --force-reinstalls "$SRC_TGZ")
+      cabal install --enable-tests --enable-benchmarks --force-reinstalls --ghc-options=-O0 --reorder-goals --max-backjumps=-1 $CABALARGS $PACKAGES
+
+      ORIGDIR=$(pwd)
+      for dir in $PACKAGES
+      do
+        cd $dir
+        cabal check || [ "$CABALVER" == "1.16" ]
+        cabal sdist
+        PKGVER=$(cabal info . | awk '{print $2;exit}')
+        SRC_TGZ=$PKGVER.tar.gz
+        cd dist
+        tar zxfv "$SRC_TGZ"
+        cd "$PKGVER"
+        cabal configure --enable-tests --ghc-options -O0
+        cabal build
+        if [ "$CABALVER" = "1.16" ] || [ "$CABALVER" = "1.18" ]; then
+          cabal test
+        else
+          cabal test --show-details=streaming --log=/dev/stdout
+        fi
+        cd $ORIGDIR
+      done
       ;;
   esac
+  set +ex

--- a/example/Client.hs
+++ b/example/Client.hs
@@ -6,14 +6,14 @@ import Data.ByteString.Lazy (ByteString)
 import qualified Data.ByteString.Lazy as LBS
 import Data.Monoid ((<>))
 import Data.Proxy (Proxy(Proxy))
-import Network.HTTP.Client (Response, defaultManagerSettings, newManager, responseBody)
-import Network.HTTP.Media (MediaType)
-import Network.HTTP.Types (Header, Method, methodGet)
+import Data.Text (Text)
+import Network.HTTP.Client (defaultManagerSettings, newManager)
+import Network.HTTP.Types (methodGet)
 import Servant.API ((:<|>)((:<|>)))
 import Servant.Client
-       (BaseUrl(BaseUrl), ClientEnv(ClientEnv), ClientM, Scheme(Http),
-        client, runClientM)
-import Servant.Common.Req (Req, appendToPath)
+       (BaseUrl(BaseUrl), ClientM, Response, Scheme(Http),
+        client, mkClientEnv, responseBody, runClientM)
+import Servant.Client.Core (Request, appendToPath, requestMethod)
 import Servant.RawM ()
 
 import Api (Api, port)
@@ -23,14 +23,16 @@ import Api (Api, port)
 -----------------------------------------
 
 otherEndpoint1 :: ClientM Int
-getFile' :: Method -> (Req -> Req) -> ClientM (Int, ByteString, MediaType, [Header], Response ByteString)
+getFile' :: (Request -> Request) -> ClientM Response
 otherEndpoint2 :: ClientM Int
 otherEndpoint1 :<|> getFile' :<|> otherEndpoint2 = client (Proxy :: Proxy Api)
 
-getFile :: String -> ClientM ByteString
+getFile :: Text -> ClientM ByteString
 getFile filePath = do
-  (_, _, _, _, resp) <- getFile' methodGet $ \req -> appendToPath filePath req
+  resp <-
+    getFile' $ \req -> appendToPath filePath req { requestMethod = methodGet }
   pure $ responseBody resp
+
 
 ----------
 -- Main --
@@ -39,7 +41,7 @@ getFile filePath = do
 main :: IO ()
 main = do
   manager <- newManager defaultManagerSettings
-  let clientEnv = ClientEnv manager baseUrl
+  let clientEnv = mkClientEnv manager baseUrl
   eitherRes <- runClientM (getFile "foo.txt") clientEnv
   case eitherRes of
     Left servantErr -> putStrLn $ "Got a ServantErr: " <> show servantErr

--- a/example/Server.hs
+++ b/example/Server.hs
@@ -8,8 +8,7 @@ import Data.Monoid ((<>))
 import Data.Proxy (Proxy(Proxy))
 import Network.Wai (Application)
 import Network.Wai.Handler.Warp (run)
-import Servant (Handler, (:<|>)((:<|>)), Server, ServerT, serve)
-import Servant.Utils.Enter ((:~>)(NT), enter)
+import Servant (Handler, (:<|>)((:<|>)), Server, ServerT, hoistServer, serve)
 
 import Servant.RawM (serveDirectoryWebApp)
 
@@ -42,14 +41,14 @@ getOtherEndpoint2 = do
   (Config _ int2 _) <- ask
   pure int2
 
+apiProxy :: Proxy Api
+apiProxy = Proxy
+
 app :: Config -> Application
-app conf = serve (Proxy :: Proxy Api) apiServer
+app conf = serve apiProxy apiServer
   where
     apiServer :: Server Api
-    apiServer = enter naturalTrans serverRoot
-
-    naturalTrans :: ReaderT Config IO :~> Handler
-    naturalTrans = NT transformation
+    apiServer = hoistServer apiProxy transformation serverRoot
 
     transformation :: ReaderT Config IO a -> Handler a
     transformation readerT = liftIO $ runReaderT readerT conf

--- a/servant-rawm.cabal
+++ b/servant-rawm.cabal
@@ -129,11 +129,13 @@ test-suite servant-rawm-test
                      , http-types
                      , servant
                      , servant-client
+                     , servant-client-core
                      , servant-rawm
                      , servant-server
                      , tasty
                      , tasty-hspec
                      , tasty-hunit
+                     , text
                      , transformers
                      , wai
                      , warp

--- a/servant-rawm.cabal
+++ b/servant-rawm.cabal
@@ -37,12 +37,12 @@ library
                      , http-types >= 0.9
                      , lens >= 4.0
                      , resourcet >= 1.0
-                     , servant-client >= 0.9
-                     , servant-docs >= 0.9
-                     , servant-server >= 0.9
+                     , servant-client >= 0.14
+                     , servant-docs >= 0.11
+                     , servant-server >= 0.14
                      , wai >= 3.2
                      , wai-app-static >= 3.1
-                     , servant-client-core
+                     , servant-client-core >= 0.14
   default-language:    Haskell2010
   ghc-options:         -Wall -fwarn-incomplete-uni-patterns -fwarn-incomplete-record-updates -fwarn-monomorphism-restriction -Wmissing-import-lists
 
@@ -61,6 +61,7 @@ executable servant-rawm-example-client
                      , servant
                      , servant-rawm
                      , servant-client
+                     , servant-client-core
                      , text
   default-language:    Haskell2010
   ghc-options:         -Wall -fwarn-incomplete-uni-patterns -fwarn-incomplete-record-updates -fwarn-monomorphism-restriction -Wmissing-import-lists -threaded -rtsopts -with-rtsopts=-N

--- a/src/Servant/RawM.hs
+++ b/src/Servant/RawM.hs
@@ -29,12 +29,15 @@ What does this look like in practice?  The following is an example of using
     filePath <- 'Control.Monad.Reader.ask'
     'serveDirectoryWebApp' filePath
 
+  apiProxy :: 'Data.Proxy.Proxy' Api
+  apiProxy = 'Data.Proxy.Proxy'
+
   app :: FilePath -> 'Network.Wai.Application'
   app filePath =
-    'Servant.Server.serve' ('Data.Proxy.Proxy' :: 'Data.Proxy.Proxy' Api) apiServer
+    'Servant.Server.serve' apiProxy apiServer
     where
-      apiServer :: Server Api
-      apiServer = 'Servant.Utils.Enter.enter' ('Servant.Utils.Enter.NT' transformation) serverRoot
+      apiServer :: 'Servant.Server.Server' Api
+      apiServer = 'Servant.Server.hoistServer' apiProxy transformation serverRoot
 
       transformation :: 'Control.Monad.Reader.ReaderT' 'FilePath' 'IO' a -> 'Servant.Server.Handler' a
       transformation readerT = 'Control.Monad.IO.Class.liftIO' $ 'Control.Monad.Reader.runReaderT' readerT filePath

--- a/src/Servant/RawM/Internal/Client.hs
+++ b/src/Servant/RawM/Internal/Client.hs
@@ -1,6 +1,7 @@
 {-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# OPTIONS_GHC -fno-warn-orphans #-}
 
@@ -20,7 +21,7 @@ This module only exports a 'HasClient' instance for 'RawM'.
 module Servant.RawM.Internal.Client where
 
 import Data.Proxy (Proxy(Proxy))
-import Servant.Client (Client, ClientM, HasClient(clientWithRoute))
+import Servant.Client (Client, HasClient(clientWithRoute, hoistClientMonad))
 import Servant.Client.Core (runRequest, RunClient, Request, Response)
 import Servant.RawM.Internal.API (RawM')
 
@@ -28,7 +29,7 @@ import Servant.RawM.Internal.API (RawM')
 --
 -- >>> :set -XTypeOperators
 -- >>> import Data.Type.Equality ((:~:)(Refl))
--- >>> Refl :: Client (RawM' a) :~: (Method -> (Request -> Request) -> ClientM (Int, ByteString, MediaType, [Header], Response ByteString))
+-- >>> Refl :: Client m (RawM' a) :~: ((Request -> Request) -> m Response)
 -- Refl
 --
 -- This allows modification of the underlying 'Request' to work for any sort of
@@ -39,15 +40,19 @@ import Servant.RawM.Internal.API (RawM')
 -- the source code repository that shows a more in-depth server, client, and
 -- documentation.
 instance RunClient m => HasClient m (RawM' serverType) where
-  type Client m (RawM' serverType) = (Request -> Request) -> ClientM Response
+  type Client m (RawM' serverType) = (Request -> Request) -> m Response
 
   clientWithRoute
     :: Proxy m
-      -> Proxy (RawM' serverType)
-      -> Request
-      -> (Request -> Request)
-      -> ClientM Response
+    -> Proxy (RawM' serverType)
+    -> Request
+    -> Client m (RawM' serverType)
+  clientWithRoute Proxy Proxy req reqFunc = runRequest $ reqFunc req
 
-  clientWithRoute Proxy Proxy r f = runRequest (f r)
-
-  -- hoistClientMethod = undefined
+  hoistClientMonad
+    :: Proxy m
+    -> Proxy (RawM' serverType)
+    -> (forall x. mon x -> mon' x)
+    -> Client mon (RawM' serverType)
+    -> Client mon' (RawM' serverType)
+  hoistClientMonad Proxy Proxy f cl = \meth -> f (cl meth)

--- a/stack.yaml
+++ b/stack.yaml
@@ -8,8 +8,7 @@ packages:
 - '.'
 
 # Packages to be pulled from upstream that are not in the resolver (e.g., acme-missiles-0.3)
-extra-deps:
-- servant-client-core-0.14.1
+extra-deps: []
 
 # Override default flag values for local packages and extra-deps
 flags: {}

--- a/test/Spec.hs
+++ b/test/Spec.hs
@@ -14,19 +14,17 @@ import Data.ByteString.Lazy (ByteString)
 import Data.Either (isLeft)
 import Data.Monoid ((<>))
 import Data.Proxy (Proxy(Proxy))
+import Data.Text (Text)
 import Data.Type.Equality ((:~:)(Refl))
 import Data.Typeable (Typeable)
-import Network.HTTP.Client
-       (Response, defaultManagerSettings, newManager, responseBody)
-import Network.HTTP.Media (MediaType)
-import Network.HTTP.Types (Header, Method, methodGet)
+import Network.HTTP.Client (defaultManagerSettings, newManager)
+import Network.HTTP.Types (methodGet)
 import Network.Wai (Application)
 import Network.Wai.Handler.Warp (run)
-import Servant ((:>), (:~>)(NT), Handler, ServerT, enter, serve)
+import Servant ((:>), Handler, ServerT, hoistServer, serve)
 import Servant.Client
-       (Client, ClientEnv(ClientEnv), ClientM, client, runClientM)
-import Servant.Common.BaseUrl (parseBaseUrl)
-import Servant.Common.Req (Req, appendToPath)
+       (Client, ClientM, Response, client, mkClientEnv, parseBaseUrl, responseBody, runClientM)
+import Servant.Client.Core (Request, appendToPath, requestMethod)
 import Test.Hspec.Wai (get, shouldRespondWith, with)
 import Test.Tasty (TestTree, defaultMain, testGroup)
 import Test.Tasty.Hspec (afterAll, beforeAll, it, shouldBe, testSpec)
@@ -88,11 +86,7 @@ checkRawMServer :: ServerT RawM m :~: m Application
 checkRawMServer = Refl
 
 checkRawMClient
-  :: Client RawM :~:
-     ( Method ->
-       (Req -> Req) ->
-       ClientM (Int, ByteString, MediaType, [Header], Response ByteString)
-     )
+  :: Client m RawM :~: ((Request -> Request) -> m Response)
 checkRawMClient = Refl
 
 instanceTests :: TestTree
@@ -118,9 +112,12 @@ fileServer = do
   serveDirectoryWebApp path
 
 app :: Application
-app = serve (Proxy :: Proxy Api) $ enter (NT trans) server
+app =
+  let proxy = Proxy :: Proxy Api
+  in
+  serve proxy $ hoistServer proxy trans server
   where
-    trans :: ReaderT FilePath IO a -> Handler a
+    trans :: forall a. ReaderT FilePath IO a -> Handler a
     trans readerT = liftIO $ runReaderT readerT "example/files"
 
 serverTestsIO :: IO TestTree
@@ -137,21 +134,21 @@ serverTestsIO =
 --------------------------------
 
 getFile'
-  :: Method
-  -> (Req -> Req)
-  -> ClientM (Int, ByteString, MediaType, [Header], Response ByteString)
+  :: (Request -> Request)
+  -> ClientM Response
 getFile' = client (Proxy :: Proxy Api)
 
-getFile :: String -> ClientM ByteString
+getFile :: Text -> ClientM ByteString
 getFile filePath = do
-  (_, _, _, _, resp) <- getFile' methodGet $ \req -> appendToPath filePath req
+  resp <-
+    getFile' $ \req -> appendToPath filePath req { requestMethod = methodGet }
   pure $ responseBody resp
 
 clientTestsIO :: IO TestTree
 clientTestsIO = do
   manager <- newManager defaultManagerSettings
   baseUrl <- parseBaseUrl $ "http://localhost:" <> show port <> "/"
-  let clientEnv = ClientEnv manager baseUrl
+  let clientEnv = mkClientEnv manager baseUrl
   testSpec "client" . beforeAll runServer . afterAll killServer $ do
     it "correctly gets files" $ \_ -> do
       eitherRes <- runClientM (getFile "bar.txt") clientEnv

